packages/cli-core/src/package-manager-controller/package_manager_controller_factory.ts

```diff
@@ -46,7 +40,11 @@ export class PackageManagerControllerFactory {
    * getPackageManagerName - returns the name of the package manager
    */
   private getPackageManagerName() {
-    const packageManagerAndVersion = this.userAgent.split(' ')[0];
+    const userAgent = process.env.npm_config_user_agent;
+    if (userAgent === undefined) {
+      throw new Error('npm_config_user_agent is undefined');
+    }
+    const packageManagerAndVersion = userAgent.split(' ')[0];
```

Comment:
Since this piece is reading from `process.env` AND it's tightly coupled to how our CLIs are integrating with package managers I think this shouldn't be called in deployer package.

After reading this PR I think we should do something like this:
1. Create plain type `type PackageManagerController` in `platform-types` and extract interface.
2. Change existing `abstract class PackageManagerController` to `abstract class PackageManagerControllerBase` and make it package private (take it out from API.md).
3. Have deployer factory accept instance of `type PackageManagerController`
4. Have package manager controller factory in cli-core
5. Use package manager controller factory in cli and pass instance to deployer factory. (so they're decoupled via plain type).

@edwardfoyle what do you think?

Reply:
1. definitely yes. I left a similar comment
2-3. this seems like a mixin pattern which I think might be overly complicating this. I think the package manager controller factory could just return subclasses of PackageManagerController without the additional abstraction. agreed that the only export should be the factory and the PackageManagerController type (not class)
4. yes
5. yes

Reply:
I mentioned 2-3 in case we want to keep shared base - there's some shared logic there afaik. It's definietly optional and since it becomes package private then it becomes lesser implementation detail.

Reply:
> edit: unless you just meant rename PackageManagerController to PackageManagerControllerBase so it doesn't conflict with the newly created PackageManagerController type

yes that's what I meant.

Reply:
1 & 2 are done

Reply:
3,4,5 are done. (I think)